Sources/ApolloCodegenLib/Frontend/JavaScriptBridge.swift

```diff
@@ -0,0 +1,345 @@
+import JavaScriptCore
+
+// JavaScriptCore APIs haven't been annotated for nullability, but most of its methods will never return `null`
+// and can be safely force unwrapped. (Even when an exception is thrown they would still return
+// a `JSValue` representing a JavaScript `undefined` value.)
+
+/// An errror thrown during JavaScript execution.
+/// See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error
+public class JavaScriptError: JavaScriptObject, Error {
+  lazy var name: String? = self["name"]
+  
+  lazy var message: String? = self["message"]
+  
+  lazy var stack: String? = self["stack"]
+}
+
+extension JavaScriptError: CustomStringConvertible {
+  public var description: String {
+    return jsValue.toString()
+  }
+}
+
+/// A type that references an underlying JavaScript object.
+public class JavaScriptObject: JavaScriptValueDecodable {
+  let jsValue: JSValue
+  unowned let bridge: JavaScriptBridge
+  
+  static func fromJSValue(_ jsValue: JSValue, bridge: JavaScriptBridge) -> Self {
+    return bridge.wrap(jsValue)
+  }
+    
+  required init(_ jsValue: JSValue, bridge: JavaScriptBridge) {
+    precondition(jsValue.isObject)
+        
+    self.jsValue = jsValue
+    self.bridge = bridge
+  }
+  
+  subscript(property: Any) -> JSValue {
+    return jsValue[property]
+  }
+  
+  subscript<Decodable: JavaScriptValueDecodable>(property: Any) -> Decodable {
+    return bridge.fromJSValue(jsValue[property])
+  }
+  
+  private func call(_ functionName: String, with arguments: [Any]) throws -> JSValue {
+    return try bridge.throwingJavaScriptErrorIfNeeded {
+      let function = jsValue[functionName]
+      
+      precondition(!function.isUndefined, "Function \(functionName) is undefined")
```

Comment:
why is this a precondition failure rather than an error? fwiw errors are a lot easier to test.

Reply:
This could conceivably be an error because it's not inside an accessor and we already rethrow errors we receive from JavaScript. But for consistency with property access, I would still argue a precondition is appropriate here because calling a non-existing function is a programmer error and not a runtime condition we can handle.

Reply:
And in theory it's a programmer error at the Apollo end, not from anything the user is doing, correct?

Reply:
Yep, that's true of all of these. It should be caught by us in development, and if it happens for users there is nothing they can do about it.